File: agent/models/trip.py
"""Pydantic stubs matching docs/DATA_MODEL.md — fill in validators as crews adopt them."""

from __future__ import annotations

from datetime import date
from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


class PlaceCategory(str, Enum):
    museum = "museum"
    food = "food"
    park = "park"
    transit = "transit"
    lodging = "lodging"
    nightlife = "nightlife"
    shopping = "shopping"
    nature = "nature"
    other = "other"


class DestinationType(str, Enum):
    city = "city"
    country = "country"
    region = "region"


class Place(BaseModel):
    place_id: Optional[str] = None
    name: str
    address: Optional[str] = None
    lat: Optional[float] = None
    lng: Optional[float] = None
    website_url: Optional[str] = None
    maps_url: Optional[str] = None
    category: PlaceCategory = PlaceCategory.other
    reason_to_visit: str = ""
    details: str = ""
    estimated_minutes: int = Field(default=60, gt=0)
    has_bathroom: Optional[bool] = None
    notes: Optional[str] = None
    order_in_day: int = 0
    place_key: str = ""


class CityStop(BaseModel):
    city: str
    country: str = ""
    nights: int = Field(default=1, ge=0)
    arrival_day_index: int = Field(default=1, ge=1)
    departure_day_index: int = Field(default=1, ge=1)
    reason: str = ""
    highlights: list[str] = Field(default_factory=list)


class CityRoute(BaseModel):
    destination_type: DestinationType
    cities: list[CityStop] = Field(default_factory=list)
    rationale: str = ""
    total_nights: int = 0
    status: str = "proposed"  # proposed | confirmed


class DayPlan(BaseModel):
    day_index: int = Field(ge=1, le=14)
    date: date
    theme: str = ""
    summary: str = ""
    overnight_city: str = ""
    places: list[Place] = Field(default_factory=list)


class Trip(BaseModel):
    trip_id: str = ""
    user_id: str = ""
    origin: str = ""
    destination: str = ""
    destination_type: DestinationType = DestinationType.city
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    day_count: int = 0
    next_day_index: int = 1
    status: str = "drafting"
    preferences: str = ""
    city_route: Optional[CityRoute] = None
    visited_place_keys: list[str] = Field(default_factory=list)
    prior_days_summary: str = ""
    days: list[DayPlan] = Field(default_factory=list)

File: frontend/src/types/trip.ts
export type DestinationType = "city" | "country" | "region";

export type CreateTripInput = {
    origin: string;
    destination: string;
    destination_type: DestinationType;
    start_date: string; // YYYY-MM-DD
    end_date: string; // YYYY-MM-DD
    preferences?: string;
};

export type Trip = {
    trip_id: string;
    origin: string;
    destination: string;
    destination_type: DestinationType;
    start_date: string; // YYYY-MM-DD
    end_date: string; // YYYY-MM-DD
    day_count: number;
    preferences?: string;
    status: string;
    next_day_index?: number;
}

export type CityStop = {
    city: string;
    country?: string;
    nights: number;
    arrival_day_index: number;
    departure_day_index: number;
    reason?: string;
    highlights?: string[];
    /** Frontend-only id so undo can target one row when city names collide. */
    client_id?: string;
};

export type Route = {
    destination_type: DestinationType;
    cities: CityStop[];
    rationale?: string;
    total_nights: number;
    status: "proposed" | "confirmed" | string;
}

export type PlaceWatchOut = {
  label: string;
  detail: string;
};

export type Place = {
  name: string;
  address?: string;
  category?: string;
  reason_to_visit?: string;
  details?: string;
  estimated_minutes?: number;
  has_bathroom?: boolean | null;
  notes?: string | null;
  order_in_day?: number;
  place_key: string;
  /** Rich detail (demo / later crew enrichment) */
  cost?: string;
  open_hours?: string;
  operational_status?: "open" | "closed" | "unknown";
  /** Monday=0 … Sunday=6 */
  closed_weekdays?: number[];
  main_attraction?: string;
  map_url?: string;
  map_embed_query?: string;
  why_suggested?: string;
  watch_outs?: PlaceWatchOut[];
  /** Minutes to reach this place from the previous stop in the day (0 for first). */
  travel_minutes_from_previous?: number;
};


export type DayPlan = {
    day_index: number;
    date: string; // YYYY-MM-DD from the API (JSON), not a Date object
    theme: string;
    summary?: string;
    overnight_city: string;
    places: Place[];
};

export type TripBundle = {
    trip: Trip;
    route: Route | null;
    days: DayPlan[];
}